### PR TITLE
Themes: Use `details` middleware for server-side rendering

### LINKED ISSUE #185: Server Side Rendering: Render a logged out layout via a pure component tree
> try and start sort of from scratch with a clean isomorphic implementation that is specific to the logged out scenario

See comment below (9287-gh-calypso-pre-oss):

> While working a bit with the developer on trial that tried to come up with something for server rendering, one idea that had was to try and start sort of from scratch with a clean isomorphic implementation that is specific to the logged out scenario. I sort of feel like implementing a bunch of workarounds is going to end up being a disservice moving forward because it will be harder for us to come up with a clean solution, and a lot of the layout related code isn't needed when logged out. That is especially true in a case like this where having good document titles and meta descriptions are valuable from an SEO perspective. 

> One big problem is that to really do server rendering we need a single React component tree. So the way that `#primary` and `#secondary` work doesn't work, which is another reason for treating the logged stuff as separate might be a good idea. We can take what was learned and then try to bring it back into the main Calypso client app.

> Have you thought much about how to rehydrate the stores on the client with data from the server? 

Note that we'll really need to benchmark any `renderToString` calls we make on the server.

### PR DESCRIPTION
**Pull request by ockham**
In the `/themes` route in `server/pages`, use `my-sites/themes` to create a theme sheet element tree in `context.layout`, and `renderToString()` that.
 * Adapt `express`'s `req` context to be able to use it with middleware that otherwise operates on `page`'s `context`. props @mcsf
 * Add a `package.json` that sets different entry points on the server and in the browser (not everything required by `my-sites/themes/index.js` is SSR-ready yet, so we need a pruned version for the server).

* Move theme details specific async data fetching to a separate middleware in  `my-sites/themes/controller/index.node.js`.
* Move generic `renderToString` caching to `server/render`.
* Cache Helmet data, so `title` etc are properly set.

To test:
* Check that everything still works as before.
* Try logged-out `/theme/mood`, with JS disabled. Note that the title is now set to '`themeName` Theme — WordPress.com' (instead of the generic 'WordPress.com', as on `master`).
* Note that logged-in `/design` and `/themes/something` routes server-render the logged-out layout (masterbar!) before it's being replaced by the logged-in one on the client. While we want to fix this in the long run, it is no behavioral change from `master`.

Note:
* There are still reconciliation errors, but there are also some in `master`. Opting to fix them later, since this PR is big enough already. (Possible cuplrit: Apparently, `react-helmet` inserts a `noscript` tag. https://github.com/nfl/react-helmet/issues/81)
* SEO is still broken for SSR'd `/design`. Also opting to fix later.

/cc @seear @ehg @mcsf

Follow-ups: #3302, #3774, #3795, #3425 

I'd say this closes #185 :-)

### CODE REVIEW

**Comment by ockham** [comment]
Getting

```
Error: Cannot find module 'indexof'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (wp-calypso/node_modules/component-classes/index.js:5:13)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
Makefile:96: recipe for target 'run' failed
make: *** [run] Error 1
```

I kinda hate `component` :-/

**Comment by ockham** [comment]
Hm, it's [this](https://github.com/Automattic/wp-calypso/blob/master/client/boot/index.js#L9) occurrence...

**Comment by ockham** [comment]
> Hm, it's this occurrence...

Argh, spoke too soon. Been on the wrong branch.

**Comment by ehg** [comment]
> Any experience from past component woos how to fix that?

NMR it, and seek an upstream fix :)

**Comment by ockham** [comment]
Been there, it seems: `8447-gh-calypso-pre-oss` :-/

**Comment by ockham** [comment]
Hmm, just adding `new webpack.NormalModuleReplacementPlugin( /^indexof$/, 'component-indexof' )` doesn't do the trick (nor does `lodash/noop`) -- probably because this is happening somewhere down in `node_modules/`, which we're not bundling through webpack?

**Comment by ockham** [comment]
> seek an upstream fix :)

component/classes#27

**Comment by ehg** [comment]
Right, so the way we're trying to avoid loading non-iso-compatible controllers is suspect. 

`import { loggedOut, details, makeLoggedOutLayout } from './controller';` will result in attempts to resolve and load all dependencies in the controller. 

What I imagined was having separate node/web controllers, so this wouldn't happen. Maybe put all SSR-compatible controllers/middleware in `controllers/index.node.js`, and have `controllers/index.web.js` specify the web-only stuff, and also pull in `controllers/index.node.js`, so we're DRY.

FWIW, I think we can solve the `component.js` issues using `webpack`'s `resolve.alias`: https://webpack.github.io/docs/configuration.html#resolve-alias — but that's not going to help us much at this point in time.

File: server/isomorphic-routing/index.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	context.initialReduxState = pick( context.store.getState(), 'ui' );
+	next();
+}
+
+function serverRender( context ) {
+	if ( config.isEnabled( 'server-side-rendering' ) ) {
+		Object.assign( context, render( context.layout ) );
+	}
+	context.res.render( 'index.jade', context );
+}
```
Comment (ockham) [inline server/isomorphic-routing/index.js]: Should move this function to `server/render`, I think.

**Comment by ockham** [comment]
Rebasing this on `master` now that #3279 is in turns out to be a major PITA. I'm afraid we'll have to untangle some code in `server/pages` with a separate PR before this is feasible...

File: server/isomorphic-routing/index.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		url: req.url,
+	} );
+}
+
+function applyMiddlewares( context, ...middlewares ) {
+	const liftedmiddlewares = middlewares.map( middleware => next => middleware( context, next ) );
+	compose( ...liftedmiddlewares )();
+}
+
+function compose( ...functions ) {
```
Comment (ockham) [inline server/isomorphic-routing/index.js]: Can we use lodash's `flowRight` instead?

File: docs/isomorphic-routing.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+For an example of isomorphic routing, see `client/my-sites/themes`.
+
+The constraints required for isomporphic routing are:
+* In `client/mysection/index.js`, export a default function that accepts
+`router` as an argument. Instead of defining routes by invoking `page`, use
+`router`, e.g.
+
+```js
+export default function( router ) {
+  router(	'/themes/:slug/:section?/:site_id?', details, makeLoggedOutLayout );
```
Comment (ehg) [inline docs/isomorphic-routing.md]: \t

File: server/pages/index.js [hunk 4]
```diff
@@ -379,95 +365,8 @@ module.exports = function() {
...
-
-	app.get( '/design(/type/:themeTier)?', function( req, res ) {
-		if ( req.cookies.wordpress_logged_in || ! config.isEnabled( 'manage/themes/logged-out' ) ) {
-			// the user is probably logged in
-			renderLoggedInRoute( req, res );
-		} else {
-			const context = getDefaultContext( req );
-			const tier = includes( [ 'all', 'free', 'premium' ], req.params.themeTier )
-				? req.params.themeTier
-				: 'all';
```
Comment (mcsf) [inline server/pages/index.js]: Not urgent, but we should still whitelist `tier` in the new implementation, especially as it's a potential vector for undesired cache manipulation.

File: client/my-sites/themes/controller/index.node.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	context.store.dispatch( setSection( 'themes', {
+		hasSidebar: false,
+		isFullScreen: true
+	} ) );
+
+	const ConnectedComponent = ( { themeSlug, contentSection } ) => (
+		<ThemeDetailsComponent id={ themeSlug } section={ contentSection } >
+			<ThemeSheetComponent />
+		</ThemeDetailsComponent>
+	);
```
Comment (mcsf) [inline client/my-sites/themes/controller/index.node.js]: We're redeclaring a component on each middleware call here.

File: client/my-sites/themes/controller/index.node.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		wpcom.undocumented().themeDetails( themeSlug, ( error, data ) => {
+			if ( error ) {
+				debug( `Error fetching theme ${ themeSlug } details: `, error.message || error );
+				return;
+			}
+			const themeData = themeDetailsCache.get( themeSlug );
+			if ( ! themeData || ( Date( data.date_updated ) > Date( themeData.date_updated ) ) ) {
+				debug( 'caching', themeSlug );
+				themeDetailsCache.set( themeSlug, data );
+				// update the render cache
```
Comment (mcsf) [inline client/my-sites/themes/controller/index.node.js]: Superfluous comment.

File: client/my-sites/themes/controller/index.node.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+				renderThemeSheet( data );
+			}
+		} );
+	}
+
+	function renderThemeSheet( theme ) {
```
Comment (mcsf) [inline client/my-sites/themes/controller/index.node.js]: Naming is hard. What if this were simply named `done` or `finish` or similar?

Comment (mcsf) [inline client/my-sites/themes/controller/index.node.js]: Or `proceed`. :]

File: client/my-sites/themes/controller/index.web.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+import MultiSiteComponent from 'my-sites/themes/multi-site';
+import LoggedOutComponent from 'my-sites/themes/logged-out';
+import analytics from 'analytics';
+import i18n from 'lib/mixins/i18n';
+import trackScrollPage from 'lib/track-scroll-page';
+import buildTitle from 'lib/screen-title/utils';
+import { getAnalyticsData } from '../helpers';
+import { setSection } from 'state/ui/actions';
+import { makeElement } from './index.node.js';
+export { details } from './index.node.js';
```
Comment (mcsf) [inline client/my-sites/themes/controller/index.web.js]: Woops, should be `import`!

Comment (mcsf) [inline client/my-sites/themes/controller/index.web.js]: Let's place this anywhere that isn't the `import` block, to make it less confusing. Maybe right before the first `export` defined in the module, like `singleSite`.

**Comment by seear** [comment]
Reviewed/tested as a team

File: server/isomorphic-routing/index.js (hunk 2 above)
Comment (mcsf) [inline server/isomorphic-routing/index.js]: (For the record, `flowRight` wasn't fit for a drop-in replacement. It was agreed that having this local implementation is fine.)

**Pull request merged**